apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/VolatileResourceManager.cs

```diff
@@ -24,46 +24,49 @@ namespace Google.Cloud.Spanner.Data
 {
     internal sealed class VolatileResourceManager : ISinglePhaseNotification, ISpannerTransaction, IDisposable
     {
-        private readonly object _lock = new object();
         private readonly SpannerConnection _spannerConnection;
         private readonly TimestampBound _timestampBound;
         private readonly TransactionId _transactionId;
-        private Task<SpannerTransaction> _transactionTask;
-        private bool _disposed = false;
+        private Lazy<Task<SpannerTransaction>> _transaction;
+        private bool _hasExecutedDml;
 
         internal VolatileResourceManager(SpannerConnection spannerConnection, TimestampBound timestampBound, TransactionId transactionId)
         {
             _spannerConnection = spannerConnection;
             _timestampBound = timestampBound;
+            _transaction = new Lazy<Task<SpannerTransaction>>(CreateTransaction, LazyThreadSafetyMode.ExecutionAndPublication);
             _transactionId = transactionId;
         }
 
+        private SpannerTransaction SpannerTransaction { get => SpannerTransactionTask.Result; }
+
+        private Task<SpannerTransaction> SpannerTransactionTask
+        {
+            get => _transaction == null ? throw new ObjectDisposedException("The transaction has already been disposed") : _transaction.Value;
+        }
+
+        private bool IsTransactionCreated { get => _transaction != null && _transaction.IsValueCreated; }
+
+        private bool HasExecutedDmlOrMutations { get => IsTransactionCreated && (_hasExecutedDml || SpannerTransaction.HasMutations); }
```

Comment:
No need for all those braces :)

```suggestion
        private SpannerTransaction SpannerTransaction => SpannerTransactionTask.Result;

        private Task<SpannerTransaction> SpannerTransactionTask =>
            _transaction == null ? throw new ObjectDisposedException("The transaction has already been disposed") : _transaction.Value;

        private bool IsTransactionCreated => _transaction != null && _transaction.IsValueCreated;

        private bool HasExecutedDmlOrMutations => IsTransactionCreated && (_hasExecutedDml || SpannerTransaction.HasMutations);
```

Reply:
Done